File: bcs-ui/backend/container_service/observability/metric/views/container.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from typing import Callable, Dict

from rest_framework.decorators import action
from rest_framework.response import Response

from backend.bcs_web.viewsets import SystemViewSet
from backend.components import bcs_monitor as prom
from backend.container_service.observability.metric.constants import (
    METRICS_DEFAULT_CONTAINER_LIST,
    METRICS_DEFAULT_NAMESPACE,
    METRICS_DEFAULT_POD_NAME,
)
from backend.container_service.observability.metric.serializers import FetchContainerMetricSLZ
from backend.utils.url_slug import URL_DEFAULT_PLACEHOLDER


class ContainerMetricViewSet(SystemViewSet):

    serializer_class = FetchContainerMetricSLZ

    def _common_query_handler(
        self, query_metric_func: Callable, cluster_id: str, pod_name: str, need_time_range: bool = True
    ) -> Dict:
        """
        查询容器指标通用逻辑

        :param query_metric_func: 指标查询方法
        :param cluster_id: 集群ID
        :param pod_name: Pod 名称
        :param need_time_range: 是否需要指定时间范围
        :return: 指标查询结果
        """
        params = self.params_validate(self.serializer_class)
        query_params = {
            'cluster_id': cluster_id,
            'namespace': params['namespace'],
            'pod_name': pod_name,
            'container_name': params.get('container_name') or ".*",
        }
        # 部分指标如 Limit 不需要时间范围
        if need_time_range:
            query_params.update(
                {
                    'start': params['start_at'],
                    'end': params['end_at'],
                }
            )

        # 添加业务ID
        query_params['bk_biz_id'] = self.request.project.cc_app_id

        return query_metric_func(**query_params)

    @action(methods=['POST'], url_path='cpu_limit', detail=False)
    def cpu_limit(self, request, project_id, cluster_id, pod_name):
        response_data = self._common_query_handler(
            prom.get_container_cpu_limit, cluster_id, pod_name, need_time_range=False
        )
        return Response(response_data)

    @action(methods=['POST'], url_path='cpu_usage', detail=False)
    def cpu_usage(self, request, project_id, cluster_id, pod_name):
        """获取指定 容器 CPU 使用情况"""
        response_data = self._common_query_handler(prom.get_container_cpu_usage_range, cluster_id, pod_name)
        return Response(response_data)

    @action(methods=['POST'], url_path='memory_limit', detail=False)
    def memory_limit(self, request, project_id, cluster_id, pod_name):
        response_data = self._common_query_handler(
            prom.get_container_memory_limit, cluster_id, pod_name, need_time_range=False
        )
        return Response(response_data)

    @action(methods=['POST'], url_path='memory_usage', detail=False)
    def memory_usage(self, request, project_id, cluster_id, pod_name):
        """获取 容器内存 使用情况"""
        response_data = self._common_query_handler(prom.get_container_memory_usage_range, cluster_id, pod_name)
        return Response(response_data)

    @action(methods=['POST'], url_path='disk_read', detail=False)
    def disk_read(self, request, project_id, cluster_id, pod_name):
        """获取 磁盘读 情况"""
        response_data = self._common_query_handler(prom.get_container_disk_read, cluster_id, pod_name)
        return Response(response_data)

    @action(methods=['POST'], url_path='disk_write', detail=False)
    def disk_write(self, request, project_id, cluster_id, pod_name):
        """获取 磁盘写 情况"""
        response_data = self._common_query_handler(prom.get_container_disk_write, cluster_id, pod_name)
        return Response(response_data)

File: bcs-services/bcs-monitor/pkg/storegw/bcs_system/source/vcluster/container.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package vcluster

import (
	"context"
	"time"

	bcsmonitor "github.com/Tencent/bk-bcs/bcs-services/bcs-monitor/pkg/component/bcs_monitor"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-monitor/pkg/storegw/bcs_system/source/base"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-monitor/pkg/utils"
	"github.com/prometheus/prometheus/prompb"
)

// handleContainerMetric 处理公共函数
func (m *VCluster) handleContainerMetric(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, promql string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {

	params := map[string]interface{}{
		"clusterID":     clusterID,
		"namespace":     namespace,
		"podname":       podname,
		"containerName": utils.StringJoinWithRegex(containerNameList, "|", "$"),
		"provider":      PROVIDER,
	}

	matrix, _, err := bcsmonitor.QueryRangeMatrix(ctx, projectID, promql, params, start, end, step)
	if err != nil {
		return nil, err
	}

	return base.MatrixToSeries(matrix), nil
}

// GetContainerCPUUsage 容器CPU使用率
func (m *VCluster) GetContainerCPUUsage(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql :=
		`sum by(container_name) (rate(container_cpu_usage_seconds_total{cluster_id="%<clusterID>s", ` +
			`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", ` +
			`%<provider>s}[2m])) * 100`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}

// GetContainerMemoryUsed 容器内存使用率
func (m *VCluster) GetContainerMemoryUsed(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql :=
		`max by(container_name) (container_memory_working_set_bytes{cluster_id="%<clusterID>s", ` +
			`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", %<provider>s})`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}

// GetContainerCPULimit 容器CPU限制
func (m *VCluster) GetContainerCPULimit(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql := `label_replace(max by(container_name) (container_spec_cpu_quota{cluster_id="%<clusterID>s", ` +
		`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", ` +
		`%<provider>s}) / 1000 > 0, "container_name", "cpu_limit", "container_name", ".*")`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}

// GetContainerMemoryLimit 容器内存限制
func (m *VCluster) GetContainerMemoryLimit(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql := `label_replace(max by(container_name) (container_spec_memory_limit_bytes{cluster_id="%<clusterID>s", ` +
		`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", %<provider>s}) > 0, ` +
		`"container_name", "memory_limit", "container_name", ".*")`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}

// GetContainerGPUMemoryUsage 容器GPU显卡使用率
func (m *VCluster) GetContainerGPUMemoryUsage(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	return nil, nil
}

// GetContainerGPUUsed 容器GPU使用量
func (m *VCluster) GetContainerGPUUsed(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	return nil, nil
}

// GetContainerGPUUsage 容器GPU使用率
func (m *VCluster) GetContainerGPUUsage(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	return nil, nil
}

// GetContainerDiskReadTotal 容器磁盘读
func (m *VCluster) GetContainerDiskReadTotal(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql :=
		`sum by(container_name) (container_fs_reads_bytes_total{cluster_id="%<clusterID>s", ` +
			`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", %<provider>s})`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}

// GetContainerDiskWriteTotal 容器磁盘写
func (m *VCluster) GetContainerDiskWriteTotal(ctx context.Context, projectID, clusterID, namespace, podname string,
	containerNameList []string, start, end time.Time, step time.Duration) ([]*prompb.TimeSeries, error) {
	promql :=
		`sum by(container_name) (container_fs_writes_bytes_total{cluster_id="%<clusterID>s", ` +
			`namespace="%<namespace>s", pod_name="%<podname>s", container_name=~"%<containerName>s", %<provider>s})`

	return m.handleContainerMetric(ctx, projectID, clusterID, namespace, podname, containerNameList, promql, start, end,
		step)
}
